leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java

```diff
@@ -47,54 +47,119 @@ public void removeListener(PresenceListener listener) {
+        return presenceStatus.isClientSleeping();
     }
 
     /**
-     * Sets the client identified by the given registration in {@link Presence#OFFLINE} mode if the current state is
-     * {@link Presence#ONLINE}. If the state is changed successfully, then the corresponding
-     * {@link PresenceListener} are notified.
+     * Set the state of the client identified by registration as {@link Presence#AWAKE}
      * 
-     * @param registration lwm2m client registration data
+     * @param registration the client's registration object
      */
-    public void setOffline(Registration registration) {
-        if (registration.usesQueueMode()
-                && transitState(registration.getEndpoint(), Presence.ONLINE, Presence.OFFLINE)) {
+    public void setAwake(Registration reg) {
+        if (reg.usesQueueMode()) {
+            clientStatusList.get(reg.getEndpoint()).setAwake();
+            startClientAwakeTimer(reg);
             for (PresenceListener listener : listeners) {
-                listener.onOffline(registration);
+                listener.onAwake(reg);
             }
         }
-
     }
 
     /**
-     * Sets the client identified by the given registration in {@link Presence.ONLINE} mode if the current state is
-     * either {@link Presence#OFFLINE} or <code>null</code>. If the state is changed successfully, then the
-     * corresponding {@link PresenceListener} are notified.
+     * Notify the listeners that the client state changed to {@link Presence#SLEEPING}. The state changes is produced
+     * inside {@link PresenceStatus} when the timer expires or when the client doesn't respond to a request.
      * 
-     * @param registration lwm2m client registration data
+     * @param registration the client's registration object
      */
-    public void setOnline(Registration registration) {
-        if (registration.usesQueueMode()
-                && transitState(registration.getEndpoint(), Presence.OFFLINE, Presence.ONLINE)) {
+    public void setSleeping(Registration reg) {
+        if (reg.usesQueueMode()) {
+            clientStatusList.get(reg.getEndpoint()).setSleeping();
+            stopClientAwakeTimer(reg);
             for (PresenceListener listener : listeners) {
-                listener.onOnline(registration);
+                listener.onSleeping(reg);
+            }
+        }
+    }
+
+    /**
+     * Creates a new {@link PresenceStatus} object associated with the client.
+     * 
+     * @param registration the client's registration object.
+     */
+    public void createPresenceStatusObject(Registration reg) {
+        clientStatusList.put(reg.getEndpoint(), new PresenceStatus());
+    }
+
+    /**
+     * Creates a new {@link PresenceStatus} object associated with the client, with a specific awake time.
+     * 
+     * @param registration the client's registration object.
+     */
+    public void createPresenceStatusObject(Registration reg, int clientAwakeTime) {

```

Comment:
I ask myself, if this is not better to create automatically the status when it does not exist on setAwake ?